### PR TITLE
[JENKINS-47896] - Introduce SerializableOnlyOverRemoting in classes.

### PR DESCRIPTION
**Pull request by oleg-nenashev**
Due to the implementation specifics, some classes in the core are serializable only over Remoting.
This change just marks these classes and utilizes the convenience method in the interface for serialization/deserialization operations. The interface is available since Remoting 3.14.

See [JENKINS-47896](https://issues.jenkins-ci.org/browse/JENKINS-47896).

No tests, because the behavior does not change.
The error propagation just becomes better in edge cases.

### Proposed changelog entries

* N/A

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@reviewbybees

### CODE REVIEW

File: core/src/main/java/hudson/FilePath.java [hunk 1]
```diff
@@ -2705,18 +2706,17 @@ public boolean isRemote() {
     }
 
     private void writeObject(ObjectOutputStream oos) throws IOException {
-        Channel target = Channel.current();
-
-        if(channel!=null && channel!=target)
+        Channel target = getChannelForSerialization();
+        if(channel!=target) {
```
Comment (jglick) [inline core/src/main/java/hudson/FilePath.java]: Note that any code which used to serialize a `FilePath` for some other reason (basic saving just `remote`) is going to start failing now. Which is probably a good thing.

(Pipeline `program.dat` uses `FilePathPickle` instead.)

**Review by jglick — APPROVED** [review]

**Review by ghost — APPROVED** [review]
🐞

File: core/src/main/java/hudson/FilePath.java (hunk 1 above)
Comment (oleg-nenashev) [reply core/src/main/java/hudson/FilePath.java]: Probably I should revert it. I need to browse the code and to remember why I did it for this class.

**Comment by oleg-nenashev** [comment]
Fixed the merge conflict after the change by @jglick in #3122

**Comment by daniel-beck** [comment]
Still (or again) conflicting.

**Comment by oleg-nenashev** [comment]
Fixed the merge conflict

File: core/src/main/java/hudson/FilePath.java (hunk 1 above)
Comment (jglick) [reply core/src/main/java/hudson/FilePath.java]: Acc. to e.g. `BuildCommandTest.sync` this is broken.

**Review by jglick — CHANGES_REQUESTED** [review]

File: core/src/main/java/hudson/FilePath.java [hunk 2]
```diff
@@ -2985,18 +2986,17 @@ public boolean isRemote() {
...
             throw new IllegalStateException("Can't send a remote FilePath to a different remote channel");
```
Comment (oleg-nenashev) [inline core/src/main/java/hudson/FilePath.java]:
```suggestion
            throw new IllegalStateException("Can't send a remote FilePath to a different remote channel (current=" + channel + ", target=" + target + ")");
```

**Comment by oleg-nenashev** [comment]
So, tests fail due to nulls: 

```
java.lang.IllegalStateException: Can't send a remote FilePath to a different remote channel (current=null, target=hudson.remoting.Channel@5c278e52:The other side of the channel)
```

File: core/src/main/java/hudson/FilePath.java [hunk 3]
```diff
@@ -2985,18 +2986,17 @@ public boolean isRemote() {
     }
 
     private void writeObject(ObjectOutputStream oos) throws IOException {
-        Channel target = Channel.current();
-
-        if(channel!=null && channel!=target)
-            throw new IllegalStateException("Can't send a remote FilePath to a different remote channel");
+        Channel target = getChannelForSerialization();
+        if (channel != null && channel != target) {
```
Comment (jglick) [inline core/src/main/java/hudson/FilePath.java]: Null check is unnecessary here—`getChannelForSerialization` already did it.

File: core/src/main/java/hudson/cli/CliManagerImpl.java [hunk 4]
```diff
@@ -47,7 +49,7 @@
  * @deprecated Specific to Remoting-based protocol.
  */
 @Deprecated
-public class CliManagerImpl implements CliEntryPoint, Serializable {
+public class CliManagerImpl implements CliEntryPoint, SerializableOnlyOverRemoting {
```
Comment (jglick) [inline core/src/main/java/hudson/cli/CliManagerImpl.java]: Do not bother, will just conflict with #3838.

File: core/src/main/java/hudson/util/StreamTaskListener.java [hunk 5]
```diff
@@ -58,7 +60,7 @@
  * 
  * @author Kohsuke Kawaguchi
  */
-public class StreamTaskListener extends AbstractTaskListener implements TaskListener, Closeable {
+public class StreamTaskListener extends AbstractTaskListener implements TaskListener, SerializableOnlyOverRemoting, Closeable {
```
Comment (jglick) [inline core/src/main/java/hudson/util/StreamTaskListener.java]: I would expect this to be marked on `TaskListener`, not `StreamTaskListener`.

**Comment by batmat** [comment]
@oleg-nenashev this PR is conflicted, are you willing to finalize it? Or you want to close it for now?
Spassiba!

**Comment by jglick** [comment]
If @oleg-nenashev does not have time, I think I could take this over—I understand the intent.

**Comment by batmat** [comment]
@jglick I guess this would be great if you can help here indeed. Given we're trying to reduce the number of open PRs, any help to push PRs to closure (be it merged or closed) is very much appreciated.

File: core/src/main/java/hudson/FilePath.java (hunk 3 above)
Comment (jglick) [reply core/src/main/java/hudson/FilePath.java]: Never mind, it is `target` which is nonnull. `channel` may be null.

**Comment by daniel-beck** [comment]
Superseded by https://github.com/jenkinsci/jenkins/pull/3980

**Comment by jglick** [comment]
Yes, though not necessary to close this one—merging the larger will automatically merge this one as well.

**Pull request closed without merging**